src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs

```diff
@@ -21,149 +28,154 @@ public sealed override string ConvertName(string name)
+                                // For example, the value 'XMLReader' should render as 'xml_reader',
+                                // however 'SHA512Hash' should render as 'sha512-hash'.
+                                if (i + 1 < chars.Length && char.IsLower(chars[i + 1]))
+                                {
+                                    WriteChar(separator, ref destination);
+                                }
+                                break;
+
+                            default:
+                                Debug.Fail($"Unexpected state {state}");
+                                break;
                         }
-                    }
 
-                    ExpandBuffer(ref result);
-                }
+                        if (lowercase)
+                        {
+                            current = char.ToLowerInvariant(current);
+                        }
 
-                if (resultUsedLength != 0)
-                {
-                    result[resultUsedLength] = _separator;
-                    resultUsedLength += 1;
-                }
+                        WriteChar(current, ref destination);
+                        state = SeparatorState.UppercaseLetter;
+                        break;
 
-                resultUsedLength += written;
-            }
+                    case UnicodeCategory.LowercaseLetter:
+                    case UnicodeCategory.DecimalDigitNumber:
 
-            int first = 0;
-            ReadOnlySpan<char> chars = name.AsSpan();
-            CharCategory previousCategory = CharCategory.Boundary;
+                        if (state is SeparatorState.SpaceSeparator)
+                        {
+                            // Normalize preceding spaces to one separator.
+                            WriteChar(separator, ref destination);
+                        }
 
-            for (int index = 0; index < chars.Length; index++)
-            {
-                char current = chars[index];
-                UnicodeCategory currentCategoryUnicode = char.GetUnicodeCategory(current);
+                        if (!lowercase)
+                        {
+                            current = char.ToUpperInvariant(current);
+                        }
 
-                if (currentCategoryUnicode == UnicodeCategory.SpaceSeparator ||
-                    currentCategoryUnicode >= UnicodeCategory.ConnectorPunctuation &&
-                    currentCategoryUnicode <= UnicodeCategory.OtherPunctuation)
-                {
-                    WriteWord(chars.Slice(first, index - first), ref result);
+                        WriteChar(current, ref destination);
+                        state = SeparatorState.LowercaseLetterOrDigit;
+                        break;
 
-                    previousCategory = CharCategory.Boundary;
-                    first = index + 1;
+                    case UnicodeCategory.SpaceSeparator:
```

Comment:
>UnicodeCategory.SpaceSeparator: [](http://example.com/codeflow?start=25&length=31)

what happen if having multiple spaces surrounded by letters? looks will be trimmed into one separator?

Reply:
Yes, this is by design. We have a few tests for that.